### PR TITLE
Zoom improvements

### LINKED ISSUE #88: Is there a way to restrict the range you can scroll to.
Is there a way to restrict the range you can scroll to. With the end and start set and with a maxZoom set you can still scroll indefinitely into the future and past.

### LINKED ISSUE #112: Disablind Drag action
Is there an option to disable the drag option if there are no event further to the left / right?

### LINKED ISSUE #168: Unable to get the zoom object
Hello everybody!

I'm currently using EventDrops in an Angular 4 application. I need to set the zoom programmatically, so I read the documentation and the example:

```
var eventDropsChart = d3.chart.eventDrops();
var element = d3.select('#chart_placeholder').datum(...);
eventDropsChart.call(eventDropsChart);
var zoom = element[0][0].zoom;
```

I have something similar in my code:

```
var chart = d3.chart.eventDrops()
            .date(d => d.date)
            .start(minDate)
            .end(maxDate)
            .mouseover(this.showTooltip)
            .mouseout(this.hideTooltip)
            .click(this.onEventClick)
            .zoomend(this.onZoomEnd)
            .eventLineColor((d, i) => colors[i]);

this.timeline = d3.select('#visual')
            .datum(data)
            .call(chart);

console.log("ZOOM");
console.log(this.timeline[0]); // Ouptut: undefined
```

but then I got undefined, so I'm unable to get the zoom.

I also tried to change the code to:

```
this.timeline = d3.select('#visual')
            .datum(data);

chart.call(chart);

console.log("ZOOM");
console.log(this.timeline[0]);
```

But then I got another error:

> Cannot read property 'each' of undefined
>     at Function.e (eventDrops.js:1)

Is it a bug or am I doing something wrong?

Thanks!

### LINKED ISSUE #236: Is it possible to dynamically add a single Drop to the timeline whenever needed?
This question is not about a bug. It's just that we need answers that we couldn't find in the documentation.

In a nutshell, we are trying to do two things:

1. Adding a single event drop to the timeline.

2. Once the event is added, we would like to programmatically scroll so that the current date and time is aligned with the right edge of the EventDrops view port.

For the moment, my main concern is the ability to add a single event.

Thanks for the great job.

### LINKED ISSUE #246: "bounds.js:10 Uncaught TypeError: Cannot read property 'length' of undefined" error
Hello,

I have this error in version 1.1.0 from https://unpkg.com/

```
bounds.js:10 Uncaught TypeError: Cannot read property 'length' of undefined
    at bounds.js:10
    at ut.call (d3.v4.min.js:2)
    at index.js:114
    at ut.call (d3.v4.min.js:2)
    at g (index.js:73)
    at ut.call (d3.v4.min.js:2)
    at init_ActivityChart (managdevices.js:740)
    at Object.complete (managdevices.js:2916)
    at i (jquery.js:2)
    at Object.fireWith (jquery.js:2)
```

How can this be solved?

Thank you!

### LINKED ISSUE #260: Enhancement: Expose zoom to allow for programatically zooming
Enhancement:

Say you have two different EventDrops charts with the same start and end dates. How would the best possible way to connect these charts be? I am looking into modifying the source (and make a PR if I figure it out) in order to do so, as I couldn't find a good way to do this initially. 

Have anyone done this? If so some help is greatly appreciated :)

### LINKED ISSUE #268: manual zoom buttons
**Problem:**
I would like to know if it is possible to create buttons that let the user zoom and scroll like a mouse? I have tried a lot of different methods and no one works. 
I really hope someone could help me, because I have spend a lot of time on it, but still run into trouble. 

**Environment**
* EventDrops version: 1.3.0
* Browser: Chrome

Thanks for your time :)

### PR DESCRIPTION
**Pull request by Busteren**
So here is a feature that has been requested a lot; zoom programmatically!

Changes:
* Added ID in the config to be able to set this on the SVG element. I found this quite useful when zooming programmatically.
* ZoomToDomain function. Takes in a domain and zooms to it. Ignores any restrictions to scale or translate (as default by D3). There are also optional arguments so it is possible to do transitions to have a neat effect. I did not manage to test the function itself, as when I tried, it says that SVGElement is unknown🤷🏼‍♂️
* Restrict panning. Default false as not to break backwards compatibility. If this is set, one cannot pan outside the start and end ranges (or what is set by default). If minimumScale on zoom is less than 1, you can zoom out, however, you won't be able to pan
* Removed redundant `new Date()` in isAfter, isBefore and withingRange.

Would be great if this PR could make it into master and then prepare this for a 1.4.0 release (as the bug fixes earlier are quite important in my opinion).

Closes #268, closes #260, closes #112, closes #88, closes #236, closes #168
Issues that were fixed in earlier PR: closes #246,

### CODE REVIEW

**Comment by jaredkirkley** [comment]
I would agree with the 1.4 release.  There are several bugs fixed in the latest master that are not in 1.3 that vastly improve the user experience.

Ones that I ran into before looking here were the color/radius not updating on zoom and the responsive width of the chart.  I could have worked around these bugs as well if I could draw manually which would also be implemented if I'm looking at this correctly.

TL;DR I'd love a 1.4 with the fixes!

**Comment by Busteren** [comment]
So the colour flickering has been fixed as you mentioned. Not sure what you mean by responsive width of the chart? There is already a method to redraw the chart, however, with zoomToDomain one can actually set [startDate, endDate] and zoom to this, which is very useful in a lot of cases. 

Think this, along with the bugfixes and vertical axis enhancement would make a great 1.4.0
Also think travis/dependencies might need an update.

**Comment by jaredkirkley** [comment]
Yes, lots of this stuff has been fixed but the fixes are not in 1.3 unless I'm mistaken.  The responsive width of the chart was fixed with https://github.com/marmelab/EventDrops/pull/249 for example, but that fix is not in 1.3 afaik.

the color/radius thing was fixed with https://github.com/marmelab/EventDrops/pull/259 but again is not in 1.3 which is the published version on npm.

I'd love to see a 1.4 with the fixes since 1.3

**Comment by Busteren** [comment]
Forgot to mention that there are some breaking changes with Travis CI, so updated it to work with the new xenial build environment. At the same time, I updated the node version to the latest LTS, as well as set Firefox to be the latest too.

**Comment by Busteren** [comment]
@jpetitcolas Hey, could you review this PR? :)

File: docs/configuration.md [hunk 1]
```diff
@@ -19,6 +19,12 @@ const chart = eventDrops({ d3 });
 
 If you use EventDrops without any module bundler, just include D3 script before EventDrops, and everything should work out of the box.
 
+## id
+
+_Default: '' (not added to svg)_
```
[the remark this answers is not in the document]
Comment (Busteren) [reply docs/configuration.md]: I agree with keeping the config simple, but I don't think I got what I was trying to achieve across. 
On zoomEnd you are able to retrieve the `d3.event` like this:
```js
        zoom: {
            onZoomEnd: () => {
                let event = d3.event;
                console.log(event.sourceEvent.target.id);
            }
        },
```
So the id is simply a way to distinguish between different charts (say two eventsDrop charts). So if you want to connect the zoom, you do it here. However, you also want to avoid an infinite loop. You could check the attributes, but they still have to be unique.
What you could do though is this:
```js
d3.event.sourceEvent.path[2].id
```
Which will get the id of the div, but that assumes that the div has an id.

What do you think? Just remove the id altogether or?

File: src/index.spec.js [hunk 2]
```diff
@@ -60,6 +76,24 @@ describe('EventDrops', () => {
         ]);
     });
 
+    // it('should zoom to correct domain on zoomToDomain', () => {
```
[the remark this answers is not in the document]
Comment (Busteren) [reply src/index.spec.js]: Will remove the code. I was trying to add a test for `zoomToDomain`, but for some reason, it would not zoom to the domain. Been trying to figure out why for some time now, but after going through the debugger I cannot see why it is not applying the zoom.

File: docs/configuration.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Busteren) [reply docs/configuration.md]: Decided to remove it. It is quite easy to add id manually: `document.querySelector('#div svg').setAttribute('id', 'svg-id')`
Might make a blog/tutorial on how to connect the zoom when I got time (probs after 1.4.0 release).

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,18 @@
 # CHANGELOG
 
+### 1.4.0 (Unreleased)
```
Comment (fzaninotto) [inline CHANGELOG.md]: please don't edit the changelog in Prs. The release manage (me) will update the changelog when we do the release.

File: CHANGELOG.md [hunk 4]
```diff
@@ -28,25 +41,25 @@
 
 ## 1.0.0
 
-*   Huge performance boost (almost 10 times faster when zooming or panning)
```
Comment (fzaninotto) [inline CHANGELOG.md]: change unrelated to the subject of your PR, please revert

File: src/isAfter.js [hunk 5]
```diff
@@ -3,6 +3,5 @@ import dateIsAfter from 'date-fns/is_after';
 export const isAfter = (date, dateBounds) => {
     const endingDate = Math.max(...dateBounds);
 
-    // @TODO: remove the `new Date()` constructor in the next major version: we need to force it at configuration level.
-    return dateIsAfter(new Date(date), endingDate);
```
Comment (fzaninotto) [inline src/isAfter.js]: this is not a new major release, please revert

File: src/isBefore.js [hunk 6]
```diff
@@ -3,6 +3,5 @@ import dateIsBefore from 'date-fns/is_before';
 export const isBefore = (date, dateBounds) => {
     const startingDate = Math.min(...dateBounds);
 
-    // @TODO: remove the `new Date()` constructor in the next major version: we need to force it at configuration level.
-    return dateIsBefore(new Date(date), startingDate);
```
Comment (fzaninotto) [inline src/isBefore.js]: please revert

File: src/withinRange.js [hunk 7]
```diff
@@ -4,6 +4,5 @@ export const withinRange = (date, dateBounds) => {
     const startingDate = Math.min(...dateBounds);
     const endingDate = Math.max(...dateBounds);
 
-    // @TODO: remove the `new Date()` constructor in the next major version: we need to force it at configuration level.
-    return isWithinRange(new Date(date), startingDate, endingDate);
+    return isWithinRange(date, startingDate, endingDate);
```
Comment (fzaninotto) [inline src/withinRange.js]: please revert that change

File: src/zoom.js [hunk 8]
```diff
@@ -15,13 +15,52 @@ export const getShiftedTransform = (
         .translate(labelsWidth + labelsPadding, 0); // put origin at its original position
 };
 
-export default (d3, svg, config, xScale, draw, getEvent) => {
+export function getDomainTransform(d3, config, zoom, domain, xScale, width) {
```
Comment (fzaninotto) [inline src/zoom.js]: please add jsDoc to explain what the function does, together with example input and output

File: CHANGELOG.md (hunk 4 above)
Comment (Busteren) [reply CHANGELOG.md]: This change was automatically done by husky, so I assume it should be there?

Comment (fzaninotto) [reply CHANGELOG.md]: you're right, let's keep your change.

File: docs/configuration.md [hunk 9]
```diff
@@ -345,7 +345,13 @@ _Default: Infinity_
 
 This parameter configures the maximum zoom level available. Set it to a lower value to prevent your users from zooming in too deeply.
 
-### numberDisplayedTicks
+### restrictpan
```
Comment (fzaninotto) [inline docs/configuration.md]: why not use camelCase like for the other parameters?

Comment (Busteren) [reply docs/configuration.md]: Ah, I had a bit of difficulty getting the GitHub anchors to work. Just tested this in another repository after looking at how GitHub does the conversion of headers to HTML and will fix later today.
It is referenced in README.md, which is why I was trying to make the anchor work (open file at the correct place) and not just open the file.

**Comment by Busteren** [comment]
Hi @fzaninotto,
I have made the changes (even though it didn't flag them as resolved for some reason.
